bundles/org.openhab.binding.thekeys/src/main/java/org/openhab/binding/thekeys/internal/api/GatewayService.java

```diff
@@ -0,0 +1,215 @@
+        return get(path, responseType, 3);
+    }
+
+    /**
+     * Make a GET request to the gateway
+     *
+     * @param path Path of the request
+     * @param responseType Response type
+     * @param remainingRetryCount Number of retry in case of request failed
+     * @return The deserialized response to the response type
+     */
+    private <T> T get(String path, Class<T> responseType, int remainingRetryCount) throws IOException {
+        try {
+            synchronized (this) {
+                String url = API_SCHEME + configuration.host + path;
+                logger.debug("Call the gateway : GET {}", url);
+                return httpClient.get(url, configuration.apiTimeout * 1000, responseType);
+            }
+        } catch (Exception e) {
+            Throwable rootCause = ExceptionUtils.getRootCause(e);
+            // Retry the same request because the gateway failed sometimes to handle the request
+            if (remainingRetryCount > 0
+                    && (rootCause instanceof ConnectException || rootCause instanceof EOFException)) {
+                logger.debug("Request failed. {}, retrying {} more times", rootCause.getClass(), remainingRetryCount);
+                return get(path, responseType, --remainingRetryCount);
+            }
+            throw e;
+        }
+    }
+
+    /**
+     * Make an authenticated POST request to the gateway
+     *
+     * @param path Path of the request
+     * @param responseType Response type
+     * @param lockId The lockId
+     * @return The deserialized response to the response type
+     */
+    private <T> T post(String path, Class<T> responseType, int lockId) throws IOException {
+        synchronized (this) {
+            String url = API_SCHEME + configuration.host + path;
+            logger.debug("Call the gateway : POST {}", url);
+            return httpClient.post(url, createRequestBody(lockId), configuration.apiTimeout * 1000, responseType);
+        }
+    }
+
+    /**
+     * Create the request body for a lock request
+     *
+     * @param lockId The lockId
+     * @return The request body as inputstream
+     */
+    private String createRequestBody(int lockId) {
+        try {
+            String timestamp = String.valueOf(Instant.now().toEpochMilli() / 1000);
+            return computeRequestBodyWithHash(timestamp, lockId);
+        } catch (Exception e) {
+            throw new TheKeysError("Failed to compute the auth data", e);
+        }
+    }
+
+    /**
+     * Compute the request body with the authenticated hash
+     *
+     * @param timestamp Timestamp of the request in seconds
+     * @param lockId The lockId
+     * @return The request body
+     */
+    String computeRequestBodyWithHash(String timestamp, int lockId) {
+        String hash = hmacSha256(timestamp, configuration.code);
+        return String.format("identifier=%s&ts=%s&hash=%s", lockId, timestamp, hash);
```

Comment:
`return "identifier=%s&ts=%s&hash=%s".formatted(lockId, timestamp, hash);`

Reply:
I didn't know this way, thanks :)